A29-xds-tls-security.md

```diff
@@ -0,0 +1,340 @@
+a Distributor is the only link between the Certificate Provider plugin code and its
+consumers which helps eliminate dependencies between the two pieces of code.
+
+A Distributor provides two interfaces:
+
+- an interface for the Certificate Provider to update certificate(s).
+
+- an interface for a consumer to register a watcher for watching certificate updates.
+
+For every incoming update from the Certificate Provider the Distributor will send out
+the same update to all its registered watchers. The Distributor can be considered as
+an extension of the Certificate Provider that allows a single Certificate Provider to
+support multiple watchers.
+
+#### gRPC Client Side Changes
+
+Let us see what changes are needed in the gRPC client architecture to use Certificate Provider
+plugins in the xDS use-case.
+
+The gRPC client xDS flow is described in
+[gRPC Client Architecture](./A27-xds-global-load-balancing.md#grpc-client-architecture).
+One of the steps in the flow involves receiving the Cluster resource in the
+[CDS response](./A27-xds-global-load-balancing.md#cds).
+When transport security (i.e. TLS/mTLS) is enabled, the response will include
+[UpstreamTlsContext](https://github.com/envoyproxy/envoy/blob/35702fed462f63a0a237cfbfdf26184272207c11/api/envoy/extensions/transport_sockets/tls/v3/tls.proto#L27)
+containing the TLS configuration needed by all the sub-channels in that cluster.
+The root-certificate information is contained in the validation context fields - either
+[validation_context_certificate_provider](https://github.com/envoyproxy/envoy/blob/35702fed462f63a0a237cfbfdf26184272207c11/api/envoy/extensions/transport_sockets/tls/v3/tls.proto#L190)
+or [combined_validation_context.validation_context_certificate_provider](https://github.com/envoyproxy/envoy/blob/35702fed462f63a0a237cfbfdf26184272207c11/api/envoy/extensions/transport_sockets/tls/v3/tls.proto#L145). For mTLS, it will also include local identity certificate
+information in [tls_certificate_certificate_provider](https://github.com/envoyproxy/envoy/blob/35702fed462f63a0a237cfbfdf26184272207c11/api/envoy/extensions/transport_sockets/tls/v3/tls.proto#L169).
+
+If a channel is using `XdsCredentials`, whenever its CDS balancer receives a
+CDS response containing UpstreamTlsContext, it will ask the
+[Certificate Provider Store](#certificate-provider-store) to find or create a
+[Certificate Provider](#certificate-provider) for each of the two `certificate_provider`
+values received. If these two values turn out to be the same, the same Certificate
+Provider will be used in both the cases. The [Distributor](#distributor)s
+associated with the Certificate Providers are passed via channel args to all the descendants
+of the CDS policy which enables a subchannel created from any of these descendant policies
+to receive Certificate updates and provide them to the associated security connector.
+
+![gRPC xDS Security Client Flow Diagram](A29_graphics/grpc_xds_security_client_flow.png)
```

Comment:
There's actually another layer of indirection here that is not captured in either the description or the diagram.  The distributor passed down into the subchannels is *not* the same one that provides data from the CertificateProvider to the CDS policy.  This layer of indirection is important for two reasons:

1. If we are using a different CertificateProvider for the root cert than for the identity cert, the CDS policy needs to merge the data from two different distributors before passing it down to the subchannels.
2. We don't want to have to recreate all subchannels whenever we get a CDS update that changes the CertificateProvider.